kuma/documenturls/models.py

```diff
@@ -0,0 +1,90 @@
+import requests
+
+from django.db import models
+from django.db.models.signals import pre_save
+from django.dispatch import receiver
+from redo import retrying
+
+
+def download_url(url, retry_options=None):
+    retry_options = retry_options or {
+        "retry_exceptions": (requests.exceptions.Timeout,),
+        "sleeptime": 2,
+        "attempts": 5,
+    }
+    with retrying(requests.get, **retry_options) as retrying_get:
+        response = retrying_get(url, allow_redirects=False)
+        response.raise_for_status()
+
+        return response
+
+
+class DocumentURL(models.Model):
+    """There are documents we look up for things like bookmarks and notes.
+    These are not legacy Wiki documents but rather remote URLs.
+    """
+
+    # E.g. /en-us/docs/web/javascript (note that it's lowercased!)
+    # (the longest URI in all our of content as of mid-2021 is 121 characters)
+    uri = models.CharField(max_length=250, unique=True, verbose_name="URI")
+    # E.g. https://developer.allizom.org/en-US/docs/Web/JavaScript
+    absolute_url = models.URLField(verbose_name="Absolute URL")
+    # If it's applicable, it's a download of the `index.json` for that URL.
+    metadata = models.JSONField(null=True)
+    # If the URI for some reason becomes invalid, rather than deleting it
+    # or storing a boolean, note *when* it became invalid.
+    invalid = models.DateTimeField(null=True)
+    created = models.DateTimeField(auto_now_add=True)
+    modified = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        verbose_name = "Document URL"
+
+    def __str__(self):
+        return self.uri
+
+    @classmethod
+    def normalize_uri(cls, uri):
+        return uri.lower().strip()
+
+
+@receiver(pre_save, sender=DocumentURL)
+def assert_lowercase_uri(sender, instance, **kwargs):
+    # Because it's so important that the `uri` is lowercased, this makes
+    # absolutely sure it's always so.
+    # Ideally, the client should check this at upsert, but if it's missed,
+    # this last resort will make sure it doesn't get in in
+    instance.uri = DocumentURL.normalize_uri(instance.uri)
+
+
+class DocumentURLCheck(models.Model):
```

Comment:
I don't see where this model is used other than the admin interface, and I don't see that its `check_uri` class method is ever used anywhere. Perhaps this is the first part of a system to be added in the future that will keep the `DocumentURL` objects up-to-date (e.g., when documents have been moved), or am I missing something?

Reply:
Exactly. It's scattered forward-thinking. I'm basically using the ORM as an audit log of the refresh/updates. Perhaps it's silly. Perhaps we can use the Python logging we could spew from our message queue workers. 

Again, not perfectly settled on how the refreshing/updating should work. Perhaps we don't bother. Like the point above, sleep on it? We could also write down these mumbling thoughts into code comments so as to remind the future self what the intention was and what dreams we had.

Reply:
Here's another thought... If we don't have a periodic job that keeps checking and updating previously saved `DocumentURL` objects, and only rely on updating (documents from the remote index.json) when there's a db write then the data can get stale until some "unlucky" user forces a refresh. 

So all in all it's not a trivial problem. Networks can be flaky and we don't want to punish an unlucky user with errors and delays, in their act of *using* their bookmark state. 

I think the more I think about it, a periodic Celery job is the best course of action. It's something we're familiar with and have routines for. 
Long story short, let's chat about it but definitely treat it as a follow-up from *this* PR.